platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,585 @@
+ */
+static int vfio_init_dev(int grp, int container,
+			 struct vfio_group_status *grp_status,
+			 struct vfio_iommu_type1_info *iommu_info,
+			 struct vfio_device_info *dev_info, char *grp_uuid)
+{
+	int device = -1;
+	int ret;
+
+	ret = ioctl(container, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU);
+	if (ret < 0) {
+		ODP_ERR("Doesn't support the IOMMU driver we want\n");
+		goto out;
+	}
+
+	/* Test the group is viable and available */
+	ret = ioctl(grp, VFIO_GROUP_GET_STATUS, grp_status);
+	if (ret < 0 || !(grp_status->flags & VFIO_GROUP_FLAGS_VIABLE)) {
+		ODP_ERR("Can't get status\n");
+		goto out;
+	}
+
+	ret = ioctl(grp, VFIO_GROUP_SET_CONTAINER, &container);
+	if (ret < 0) {
+		ODP_ERR("Failed to set container\n");
+		goto out;
+	}
+
+	ret = ioctl(container, VFIO_SET_IOMMU, VFIO_TYPE1_IOMMU);
+	if (ret < 0) {
+		ODP_ERR("Failed to set IOMMU\n");
+		goto out;
+	}
+
+	ret = ioctl(container, VFIO_IOMMU_GET_INFO, iommu_info);
+	if (ret < 0) {
+		ODP_ERR("Failed to get IOMMU info\n");
+		goto out;
+	}
+
+	ODP_DBG("iova_pgsizes bitmask=0x%llx\n", iommu_info->iova_pgsizes);
+	/* Get a file descriptor for the device */
+	device = ioctl(grp, VFIO_GROUP_GET_DEVICE_FD, grp_uuid);
+	if (device < 0) {
+		ODP_ERR("Failed to get device FD\n");
+		goto out;
+	}
+
+	/* Test and setup the device */
+	ret = ioctl(device, VFIO_DEVICE_GET_INFO, dev_info);
+	if (ret < 0) {
+		ODP_ERR("Failed to get device info\n");
+		goto out;
+	}
+
+	ODP_DBG("Device %d Regions: %d, irqs:%d\n", device,
+		dev_info->num_regions, dev_info->num_irqs);
+
+out:
+	return device;
+}
+
+int mdev_device_create(mdev_device_t *mdev, const char *mod_name,
+		       const char *if_name,
+		       mdev_region_info_cb_t region_info_cb)
+{
+	struct vfio_group_status group_status = {
+		.argsz = sizeof(group_status)
+	};
+	struct vfio_iommu_type1_info iommu_info = {
+		.argsz = sizeof(iommu_info)
+	};
+	struct vfio_device_info device_info = {
+		.argsz = sizeof(device_info)
+	};
+	int ret;
+
+	memset(mdev, 0, sizeof(*mdev));
+	mdev->container = -1;
+	mdev->group = -1;
+
+	strncpy(mdev->if_name, if_name, sizeof(mdev->if_name) - 1);
+
+	mdev->group_id =
+	    mdev_sysfs_discover(mod_name, mdev->if_name, mdev->group_uuid,
+				sizeof(mdev->group_uuid));
+	if (mdev->group_id < 0)
+		goto fail;
+
+	mdev->container = get_container();
+	if (mdev->container < 0)
+		goto fail;
+
+	mdev->group = get_group(mdev->group_id);
+	if (mdev->group < 0)
+		goto fail;
+
+	mdev->device =
+	    vfio_init_dev(mdev->group, mdev->container, &group_status,
+			  &iommu_info, &device_info, mdev->group_uuid);
+	if (mdev->device < 0)
+		goto fail;
+
+	ret = -EINVAL;
+	for (uint32_t region = 0; region < device_info.num_regions; region++) {
+		struct vfio_region_info *region_info;
+
+		region_info = vfio_get_region(mdev, region);
+		if (!region_info)
+			continue;
+
+		ret = region_info_cb(mdev, region_info);
+		if (ret < 0) {
+			ODP_ERR("Region info cb fail on region_info[%u]\n",
+				region);
+			return -1;
```

Comment:
Doesn't this return leak `region_info` here? You free it only after this check.

Reply:
Thanks, fixed.